### PR TITLE
User controller

### LINKED ISSUE #45: User Controller
- [x] AddUser `users`
- [x] GetUser `users/{id}`
- [x] GetBooks `users/{id}/books`
- [x] UpdateUser `users/{id}`
- [x] DeleteUser `users/{id}`
- [x] SuspendUser `users/{id}/suspend`

### PR DESCRIPTION
**Pull request by Sjimooon**
- Added `UserController`
- Added `PubHub.Common` assembly with models.

Closes #45.

### CODE REVIEW

File: PubHub/PubHub.API/Controllers/UserController.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        public IActionResult GetUser(int id)
+        {
+            var userModel = _context.PubHubUsers
+                .Include(u => u.Account)
+                .ThenInclude(a => a!.AccountType)
+                .Select(u => u.MapUserModelProperties())
+                .SingleOrDefault(u => u.Id == id);
+            if (userModel == null)
+            {
+                return NotFound($"No user with ID: {id}");
```
Comment (ZhakalenDk) [inline PubHub/PubHub.API/Controllers/UserController.cs:35]: Should we not use [RFC9110](https://datatracker.ietf.org/doc/html/rfc9110) as error response standard?

**Review by ZhakalenDk — CHANGES_REQUESTED** [review]
You need to change the stuff we discussed

**Review by jeongoks — APPROVED** [review]

File: PubHub/PubHub.API/Domain/Entities/User.cs [hunk 2]
```diff
@@ -5,13 +5,13 @@ namespace PubHub.API.Domain.Entities
...
         public int Id { get; set; }
-        public int? AccountId { get; set; }
+        public int AccountId { get; set; }
         public required string Name { get; set; }
         public required string Surname { get; set; }
         public required DateOnly Birthday { get; set; }
 
         #region Navs
-        public Account? Account { get; set; }
+        public Account Account { get; set; } = default!;
```
Comment (ZhakalenDk) [inline PubHub/PubHub.API/Domain/Entities/User.cs:14]: Should be `null` instead of `default`

File: PubHub/PubHub.API/Domain/PubHubContext.cs [hunk 3]
```diff
@@ -4,11 +4,10 @@
 using PubHub.API.Domain.Entities;
 using PubHub.API.Domain.Extensions;
 using PubHub.API.Domain.Identity;
-using System.Reflection.Emit;
 
 namespace PubHub.API.Domain
 {
-    public sealed class PubHubContext : IdentityDbContext<Account, IdentityRole<int>, int>
+    public class PubHubContext : IdentityDbContext<Account, IdentityRole<int>, int>
```
Comment (ZhakalenDk) [inline PubHub/PubHub.API/Domain/PubHubContext.cs:10]: Why did you remove `sealed`?

File: PubHub/PubHub.Common/Models/BookModel.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿namespace PubHub.Common.Models
+{
+    public class BookModel
```
Comment (ZhakalenDk) [inline PubHub/PubHub.Common/Models/BookModel.cs:3]: As discussed the naming convetion for models are: [Entity][Function]Model

Example: UserUpdateModel

**Review by ZhakalenDk — CHANGES_REQUESTED** [review]

File: PubHub/PubHub.API/Controllers/UserController.cs [hunk 5]
```diff
@@ -119,7 +120,7 @@ public async Task<IActionResult> UpdateUserAsync(int id, [FromBody] UserUpdateMo
             _context.Set<User>().Update(user);
             await _context.SaveChangesAsync();
```
Comment (ZhakalenDk) [inline PubHub/PubHub.API/Controllers/UserController.cs:121]: I would confirm that the process has actually been perfomed.
Like: 

```cs
if (!(await _context.SaveChangesAsync() > 0))
    return Results.Problem (/*Bla bla*/);
```

File: PubHub/PubHub.API/Controllers/UserController.cs [hunk 6]
```diff
@@ -142,23 +143,91 @@ public async Task<IActionResult> DeleteUserAsync(int id)
...
+            Dictionary<string, object?> extensions = new()
+            {
+                ["UserId"] = id,
+                ["AccountId"] = accountId
+            };
+
+            return Results.Problem(
+                statusCode: StatusCodes.Status500InternalServerError,
+                detail: $"Unable to delete account (ID: {accountId}) of user (ID: {id}).",
+                extensions: extensions);
```
Comment (ZhakalenDk) [inline PubHub/PubHub.API/Controllers/UserController.cs:170]: Is extensions the same as instance, or is my understanding flawed here?

File: PubHub/PubHub.API/Controllers/UserController.cs [hunk 7]
```diff
@@ -142,23 +143,91 @@ public async Task<IActionResult> DeleteUserAsync(int id)
...
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public async Task<IResult> SuspendUserAsync(int id)
+        {
+            const string accountTypeName = "suspended";
+
+            // Get account type ID.
+            var accountTypeId = await _context.Set<AccountType>()
+                .FirstOrDefaultAsync(a => a.Name.Equals(accountTypeName, StringComparison.InvariantCultureIgnoreCase));
+            if (accountTypeId == default)
```
Comment (ZhakalenDk) [inline PubHub/PubHub.API/Controllers/UserController.cs:191]: Again `default` should not be used in this context. It's difficult to know what the default value is wihtout actually looking at the type.

I suggest cheching for `null` if that is the intend.

File: PubHub/PubHub.API/Controllers/UserController.cs [hunk 8]
```diff
@@ -142,23 +143,91 @@ public async Task<IActionResult> DeleteUserAsync(int id)
...
+                    statusCode: StatusCodes.Status500InternalServerError,
+                    detail: $"Unable to find account type: '{accountTypeName}'");
+            }
+
+            // Get account ID.
+            var accountId = await _context.Set<User>()
+                .Where(u => u.Id == id)
+                .Select(u => u.AccountId)
+                .SingleOrDefaultAsync();
+            if (accountId == default)
```
Comment (ZhakalenDk) [inline PubHub/PubHub.API/Controllers/UserController.cs:203]: Again; I suggest checking for `null` if that is the intend.

**Review by ZhakalenDk — CHANGES_REQUESTED** [review]
I have some minor issues with some of the code.

Genrally when using `default` we do it when we don't know the default value.
A good example is with generics, as we can't check for the default without knowing the actual type, or the constraint of it.

```cs
public T CreateInstace<T> ()
{
    T instance = default;
    return instance;
}
```

File: PubHub/PubHub.API/Controllers/UserController.cs (hunk 6 above)
Comment (Sjimooon) [reply PubHub/PubHub.API/Controllers/UserController.cs:170]: I think they are different. Both `instance` and `extensions` can be set at the same time and `instance` is of type `string?`, whereas `extensions` is a dictionary. 🤔

File: PubHub/PubHub.API/Controllers/UserController.cs (hunk 7 above)
Comment (Sjimooon) [reply PubHub/PubHub.API/Controllers/UserController.cs:197]: I discovered that the `accountTypeId` wasn't actually the ID but the whole entity. It should be fixed to be only the ID now.

Comment (ZhakalenDk) [reply PubHub/PubHub.API/Controllers/UserController.cs:197]: > We cannot check for null as the type is `int`. Should we just check with `accountTypeId == 0` instead? I thought using `default` would make it more readable when reading it like so:
> 
> 1. `accountTypeId` is first or default.
> 2. If `accountTypeId` is default; do this.

Aah, I see. I thought the ID was an `int?`.
I'd actually rather have it be specific, but I don't know. It's just my opinion.
@jeongoks What's your take on this?

I could go both ways tbh. I'd just prefer it if we're explicit where we can be 🤔

File: PubHub/PubHub.API/Controllers/UserController.cs (hunk 6 above)
Comment (ZhakalenDk) [reply PubHub/PubHub.API/Controllers/UserController.cs:174]: Interesting. Might need to read up on th especification 👌

File: PubHub/PubHub.API/Controllers/UserController.cs (hunk 7 above)
Comment (Sjimooon) [reply PubHub/PubHub.API/Controllers/UserController.cs:197]: Shouldn't we just go with 0 then? I will be doing something similar in the PublisherController but with `-1` as a second result value. Using `0` instead of `default` there, will make it look more pretty 🙃

Comment (ZhakalenDk) [reply PubHub/PubHub.API/Controllers/UserController.cs:197]: I would prob. do something like `const int INVALID_ID = 0` as that's more readable.
And I would do so for -1 as well in relation to whatever you're using it for 🤔

Comment (Sjimooon) [reply PubHub/PubHub.API/Controllers/UserController.cs:197]: Yes, I agree with that solution. 🙂 I will define some local constants to make the code more readable.

**Review by ZhakalenDk — APPROVED** [review]

File: PubHub/PubHub.API/Controllers/UserController.cs (hunk 7 above)
Comment (jeongoks) [reply PubHub/PubHub.API/Controllers/UserController.cs]: Seems like you guys came to a conclusion on your own.

However, I do also like being more specific when we're doing checks.
